blocks/search.py

```diff
@@ -0,0 +1,401 @@
+        Sequence generator to be used
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, sequence_generator):
+        self.generator = sequence_generator
+        self.compiled = False
+
+    @abstractmethod
+    def compile(self):
+        self.compiled = True
+
+    @abstractmethod
+    def search(self, **kwargs):
+        r"""Performs search.
+
+        Parameters
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+
+        """
+        if not self.compiled:
+            self.compile()
+
+
+def unchunk_rename(*args, **kwargs):
+    def _trace(func):
+        def wrapper(self, *inputs):
+            merged_input = []
+            for input in inputs:
+                merged_input += [OrderedDict([(name, self.merge_chunks(value))
+                                              for name, value
+                                              in input.iteritems()])]
+            names, outputs = func(self, *merged_input)
+            if add_time_dim:
+                outputs = [self.divide_by_chunks(value.reshape((1,) +
+                                                               value.shape))
+                           for value in outputs]
+            else:
+                outputs = [self.divide_by_chunks(value.reshape(value.shape))
+                           for value in outputs]
+            return OrderedDict(zip(names, outputs))
+        return wrapper
+    if len(args) == 1 and callable(args[0]):
+        # No arguments, this is the decorator
+        # Set default values for the arguments
+        add_time_dim = True
+        return _trace(args[0])
+    else:
+        # This is just returning the decorator
+        add_time_dim = kwargs['add_time_dim']
+        return _trace
+
+
+class BeamSearch(Search):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    batch_size : int
+        Size of input batch.
+    sequence_generator : sequence generator
+        Sequence generator brick.
+    attended : theano variable
+        Theano variable for attended in sequence generator.
+    attended_mask : theano variable
+        Theano variable for attended mask in sequence generator.
+    inputs_dict : dict
+        Dictionary of inputs {name: theano variable for input}. The
+        functions will be constructed with these inputs.
+
+    """
+    def __init__(self, beam_size, batch_size, sequence_generator, attended,
+                 attended_mask, inputs_dict):
+        super(BeamSearch, self).__init__(sequence_generator)
+        self.beam_size = beam_size
+        self.batch_size = batch_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.attended_mask = attended_mask
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate_outputs()
```

Comment:
I was wrong that it can't work, but please use `generate.output` instead. The functions that implement logics for properties were not meant to be invoked explicitly.